src/api/uploads/BaseUpload.js

```diff
@@ -60,11 +69,51 @@ class BaseUpload extends Base {
         this.xhr.options({
             url,
             data: attributes,
-            successHandler: this.preflightSuccessHandler,
+            successHandler: this.preflightSuccessReachabilityHandler,
             errorHandler: this.preflightErrorHandler,
         });
     };
 
+    /**
+     * Handles successful preflight response.
+     * Performs a upload reachability test before calling preflightSuccessHandler.
+     *
+     * @param {Object} - Request options
+     * @return {Promise} Async function promise
+     */
+    preflightSuccessReachabilityHandler = async ({ data }: { data: { upload_url?: string } }): Promise<any> => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const uploadUrl = data.upload_url;
+        // If uploadUrl is not avaliable, don't make reachability test
+        if (!uploadUrl) {
+            this.preflightSuccessHandler({ data });
+            return;
+        }
+
+        // Parse upload host from uploadUrl
+        const splitUrl = uploadUrl.split('/');
+        const uploadHost = `${splitUrl[0]}//${splitUrl[2]}/`;
+        // The default upload host should always reachable
+        if (uploadHost === `${DEFAULT_HOSTNAME_UPLOAD}/`) {
+            this.preflightSuccessHandler({ data });
+            return;
+        }
+
+        // If upload host reachable upload file, else make a new preflight request
+        const isHostReachable = await this.uploadsReachability.isReachable(uploadHost);
+        if (isHostReachable) {
+            this.preflightSuccessHandler({ data });
+        } else if (this.reachabilityRetryCount >= MAX_REACHABILITY_RETRY) {
```

Comment:
if we hit the max retries, should the host be considered unreachable?

also correct me if i'm wrong, but we want to implement retry logic in reaching the upload host not retry the preflight?

Reply:
In this case we would only hit this `else if`  when both the reachabilityRetryCount is greaterThenOrEqual to the max retry count and when the reachability test fails. Given that it doesn't make sense to pass in `data` to the preflightSuccessHandler since we already know that host is unreachable and the upload will just fail.

If the reachability test fails for a host, it doesn't make sense to retry the reachability test again since it will just fail again. Once we know a host is not reachable, in order to get a new host we need to do a preflight check again but this time with the `unreachable_hosts` query parameter.